### PR TITLE
Add highlight color and show in mentions to automod messages

### PR DESCRIPTION
**Pull request by KleberPF**
For now, the added/removed term messages don't show up in /mentions, but I'll add a commit doing that if we decide to do it.

## Previous messages change color when the user changes settings
image
image

## Scrollbar color
image

## Scrollbar color also changes when settings are changed
image

## Message showing up in mentions
image

## Message doesn't show up in mentions if setting is disabled
image
image

I was hitting an assert when allowing/denying a message, but I tested in master and it also happens there. It seems to be caused by the request created in `Helix::manageAutoModMessages` not being concurrent, but I didn't look too much into it. This was the call stack when the assert triggered

image

Fixes: #5206

### CODE REVIEW

**Comment by Nerixyz** [comment]
You didn't cause the assertion issue: `isInStreamerMode` is called in a PubSub handler: https://github.com/Chatterino/chatterino2/blob/ea19c5c989dbea36b114bd524d50457adcb27206/src/Application.cpp#L710
The PubSub handler is running in it's own thread - not the GUI thread. `isInStreamerMode` isn't annotated, but has to run on a single thread as it accesses `cache`. This thread must be the GUI thread as it emits `streamerModeChanged`.

**Comment by KleberPF** [comment]
Oh, nice. I think I mixed some things up. I put a breakpoint in `isGuiThread` for when the result was false, but this function isn't called just inside `assertInGuiThread`, and it ends up begin called by some functions in `NetworkData`. Ended up posting the correct call stack

File: src/messages/layouts/MessageLayout.cpp [hunk 1]
```diff
@@ -376,11 +376,22 @@ void MessageLayout::updateBuffer(QPixmap *buffer,
              this->message_->flags.has(MessageFlag::LowTrustUsers))
     {
         backgroundColor = QColor("#404040");
+        if (this->message_->flags.has(MessageFlag::AutoModOffendingMessage) &&
+            ctx.preferences.enableAutomodHighlight)
+        {
+            backgroundColor =
+                *ctx.colorProvider.color(ColorType::AutomodHighlight);
+        }
```
Comment (KleberPF) [inline src/messages/layouts/MessageLayout.cpp:384]: Not sure if I should blend colors here

File: src/messages/layouts/MessageLayout.cpp [hunk 2]
```diff
@@ -376,11 +376,22 @@ void MessageLayout::updateBuffer(QPixmap *buffer,
...
     }
     else if (this->message_->flags.has(MessageFlag::Debug))
     {
         backgroundColor = QColor("#4A273D");
     }
+    else if (this->message_->flags.has(MessageFlag::AutoModOffendingMessage) &&
+             ctx.preferences.enableAutomodHighlight)
+    {
+        backgroundColor = *ctx.colorProvider.color(ColorType::AutomodHighlight);
+    }
```
Comment (KleberPF) [inline src/messages/layouts/MessageLayout.cpp:394]: Forgot to remove, commenting to remind me

File: src/messages/layouts/MessageLayout.cpp (hunk 1 above)
Comment (Nerixyz) [reply src/messages/layouts/MessageLayout.cpp]: Yes, they should be blended, but with the original background color - not the automod color. Previously, they weren't blended, as `#404040` is fully opaque. Now, users can set a color with transparency which would blend with the default color of a `QPixmap` - the backing buffer for a message (that one is actually undefined but usually some light gray with MSVC).

You could put the assignment to the default color in the `else` branch.

Comment (KleberPF) [reply src/messages/layouts/MessageLayout.cpp]: Got it, I think it's fixed now.

**Review by pajlada — COMMENTED** [review]
I think the "header", i.e. message that contains the Allow & Deny buttons, should be colored based on the AutomodHighlight

File: src/providers/colors/ColorProvider.cpp [hunk 3]
```diff
@@ -128,6 +128,9 @@ void ColorProvider::initTypeColorMap()
     initColor(ColorType::ThreadMessageHighlight,
               getSettings()->threadHighlightColor,
               HighlightPhrase::FALLBACK_THREAD_HIGHLIGHT_COLOR);
+
+    initColor(ColorType::AutomodHighlight, getSettings()->automodHighlightColor,
+              HighlightPhrase::FALLBACK_HIGHLIGHT_COLOR);
```
Comment (pajlada) [inline src/providers/colors/ColorProvider.cpp:133]: This should probably have its own fallback color `#404040`, which was the default automod highlight color before

**Comment by KleberPF** [comment]
> I think the "header", i.e. message that contains the Allow & Deny buttons, should be colored based on the AutomodHighlight

Is it ok to set `MessageFlag::AutoModOffendingMessage` for that message or should I make a new flag?

File: src/providers/colors/ColorProvider.cpp (hunk 3 above)
Comment (KleberPF) [reply src/providers/colors/ColorProvider.cpp]: Added `HighlightPhrase::FALLBACK_AUTOMOD_HIGHLIGHT_COLOR` with color `QColor(64, 64, 64);`

**Comment by Nerixyz** [comment]
> Is it ok to set `MessageFlag::AutoModOffendingMessage` for that message or should I make a new flag?

The issue with reusing the flag is that you'd create two highlights. You can avoid creating a new flag by checking for `MessageFlag::Timeout`, which is only set for offending messages, and the "header" in combination with `AutoMod`.

**Comment by KleberPF** [comment]
> The issue with reusing the flag is that you'd create two highlights. You can avoid creating a new flag by checking for `MessageFlag::Timeout`, which is only set for offending messages, and the "header" in combination with `AutoMod`.

Something like this?

```
if (ctx.preferences.enableAutomodHighlight && 
            (this->message_->flags.has(MessageFlag::AutoModOffendingMessage) || 
             this->message_->flags.has(MessageFlag::Timeout))
```

> checking for `MessageFlag::Timeout`, which is only set for offending messages

Is it a good idea to rely on this? I feel like it could break something in the future

**Comment by Nerixyz** [comment]
> Is it a good idea to rely on this? I feel like it could break something in the future (I guess it's unlikely because we would the same combination of flags)

Yea, it's probably better to have a new flag.

**Comment by KleberPF** [comment]
Added coloring for the message header
image
in mentions
image

**Comment by Mm2PL** [comment]
Is it intended that AutoMod's "Hey! Your message is being checked" is not customizable with your changes?

**Comment by KleberPF** [comment]
> Is it intended that AutoMod's "Hey! Your message is being checked" is not customizable with your changes?

Good question. I interpreted the issue as "highlight messages captured by automod", not "highlight that automod caught your message", but I'm not sure about the original intent.

**Comment by Mm2PL** [comment]
They look really similar right now on `master` and are from the same source. I am of the opinion that they should be thrown into the same bag as mod-side automod messages. I don't think the issue author even considered their existence. Interestingly enough currently these messages don't show up in `/automod`.

**Comment by pajlada** [comment]
I interpreted this as a moderator/broadcaster tool, so that's what I think is important

**Review by Mm2PL — APPROVED** [review]

**Review by pajlada — APPROVED** [review]

**Review by pajlada — APPROVED** [review]
Thank you @KleberPF !
Please don't forget to make a separate PR adding yourself into the About page :)

**Pull request merged**